File: tools/dos_hook_test.py
#!/usr/bin/env python3
"""Regression tests for tools/dos_hook.py -- the versioned dos-hook launcher (#2705).

The load-bearing witness (issue #2705 acceptance):
  * ``test_argparse_error_does_not_block`` -- a **simulated dos argparse error**
    (the Python fallback exits code 2, Claude Code's *block* code, with no deny on
    stdout) must NOT block the tool call: ``main`` exits **0** and emits no deny. This
    is the version-drift hard-block the issue exists to kill.

Plus the surrounding contract:
  * a real deny (``permissionDecision: deny`` JSON on stdout) is PRESERVED whether the
    native binary or the Python fallback emits it (exit coerced to 0, deny unaffected);
  * native OWNED (rc == 0) short-circuits -- the Python path is not consulted;
  * native DELEGATE (rc != 0) / **binary absent** falls back to Python, with the SAME
    buffered stdin (a native->python delegation re-decides on the real payload);
  * the launcher never raises -- an internal error still exits 0 (fail-safe).
"""
from __future__ import annotations

import unittest
from pathlib import Path

import dos_hook


class FakeRunner:
    """Records every (argv, stdin) it is handed and replies from a scripted queue.

    Each scripted reply is ``(returncode, stdout_bytes)``. Calls beyond the script
    reuse the last reply (so a test can script just the branch it cares about)."""

    def __init__(self, replies):
        self.replies = list(replies)
        self.calls = []  # list of (argv, stdin_bytes)

    def __call__(self, argv, stdin_data):
        self.calls.append((list(argv), stdin_data))
        idx = min(len(self.calls) - 1, len(self.replies) - 1)
        return self.replies[idx]


DENY_JSON = b'{"hookSpecificOutput":{"permissionDecision":"deny"}}'
PAYLOAD = b'{"tool_name":"Bash","tool_input":{"command":"rm -rf /"}}'


class RunHookContract(unittest.TestCase):
    def test_argparse_error_does_not_block(self):
        # THE witness: native absent, Python fallback simulates an argparse error
        # (exit code 2, no stdout). The tool call must NOT be blocked.
        runner = FakeRunner([(2, b"")])
        out = dos_hook.run_hook("pretool", ".", PAYLOAD, native=None, run=runner)
        self.assertEqual(out, b"")  # no deny emitted -> call proceeds
        # and end-to-end the launcher process exits 0 despite the code-2 fallback.
        rc = dos_hook.main(
            argv=["pretool", "--workspace", "."],
            stdin_bytes=PAYLOAD,
            run=FakeRunner([(2, b"")]),
            env={"CLAUDE_PROJECT_DIR": "/no/such/root"},  # forces binary-absent
        )
        self.assertEqual(rc, 0)

    def test_real_deny_from_python_is_preserved(self):
        # A genuine deny on stdout survives even though the fallback also exits non-zero.
        runner = FakeRunner([(2, DENY_JSON)])
        out = dos_hook.run_hook("pretool", ".", PAYLOAD, native=None, run=runner)
        self.assertEqual(out, DENY_JSON)

    def test_native_owned_short_circuits(self):
        # rc == 0 = OWNED: forward native stdout, do NOT consult Python.
        native = Path("tools/.bin/dos-hook")
        runner = FakeRunner([(0, DENY_JSON)])
        out = dos_hook.run_hook("pretool", ".", PAYLOAD, native=native, run=runner)
        self.assertEqual(out, DENY_JSON)
        self.assertEqual(len(runner.calls), 1)  # native only
        self.assertIn("dos-hook", runner.calls[0][0][0])

    def test_native_delegate_falls_back_to_python_same_stdin(self):
        # rc != 0 = DELEGATE: fall to Python, re-deciding on the SAME buffered stdin.
        native = Path("tools/.bin/dos-hook")
        runner = FakeRunner([(3, b""), (0, DENY_JSON)])
        out = dos_hook.run_hook("posttool", ".", PAYLOAD, native=native, run=runner)
        self.assertEqual(out, DENY_JSON)
        self.assertEqual(len(runner.calls), 2)  # native then python
        # both backends saw the identical, undrained payload
        self.assertEqual(runner.calls[0][1], PAYLOAD)
        self.assertEqual(runner.calls[1][1], PAYLOAD)
        # second call is the python -m dos.cli fallback
        self.assertIn("dos.cli", " ".join(runner.calls[1][0]))

    def test_binary_absent_uses_python(self):
        runner = FakeRunner([(0, b"")])
        dos_hook.run_hook("pretool", ".", PAYLOAD, native=None, run=runner)
        self.assertEqual(len(runner.calls), 1)
        self.assertIn("dos.cli", " ".join(runner.calls[0][0]))


class LauncherFailSafe(unittest.TestCase):
    def test_main_swallows_runner_exception(self):
        def boom(argv, stdin_data):
            raise RuntimeError("backend blew up")

        rc = dos_hook.main(
            argv=["pretool", "--workspace", "."],
            stdin_bytes=PAYLOAD,
            run=boom,
            env={"CLAUDE_PROJECT_DIR": "/no/such/root"},
        )
        self.assertEqual(rc, 0)  # a launcher/backend error must never wedge the fleet

    def test_native_binary_resolution_prefers_project_dir(self):
        # Absent binary under a bogus root resolves to None (Python fallback).
        self.assertIsNone(dos_hook.native_binary(Path("/no/such/root")))

    def test_repo_root_prefers_claude_project_dir(self):
        root = dos_hook.repo_root(env={"CLAUDE_PROJECT_DIR": "/some/ws"})
        self.assertEqual(str(root).replace("\\", "/"), "/some/ws")


if __name__ == "__main__":
    unittest.main()

File: internal/doshook/doshook_test.go
package doshook

import (
	"bytes"
	"errors"
	"io"
	"path/filepath"
	"strings"
	"testing"
)

type reply struct {
	exitCode int
	stdout   []byte
	err      error
}

type callRecord struct {
	argv  []string
	stdin []byte
}

type fakeRunner struct {
	replies []reply
	calls   []callRecord
}

func newFakeRunner(replies ...reply) *fakeRunner {
	return &fakeRunner{
		replies: replies,
	}
}

func (f *fakeRunner) Run(argv []string, stdin []byte) (int, []byte, error) {
	argvCopy := make([]string, len(argv))
	copy(argvCopy, argv)
	stdinCopy := make([]byte, len(stdin))
	copy(stdinCopy, stdin)
	f.calls = append(f.calls, callRecord{argv: argvCopy, stdin: stdinCopy})

	if len(f.replies) == 0 {
		return 0, nil, nil
	}
	idx := len(f.calls) - 1
	if idx >= len(f.replies) {
		idx = len(f.replies) - 1
	}
	r := f.replies[idx]
	return r.exitCode, r.stdout, r.err
}

var (
	denyJSON    = []byte(`{"hookSpecificOutput":{"permissionDecision":"deny"}}`)
	testPayload = []byte(`{"tool_name":"Bash","tool_input":{"command":"rm -rf /"}}`)
)

func TestNativeOwnedReturnsNativeStdout(t *testing.T) {
	native := filepath.Join("tools", ".bin", "dos-hook")
	runner := newFakeRunner(reply{exitCode: 0, stdout: denyJSON})
	out, err := RunHook("pretool", ".", testPayload, native, runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !bytes.Equal(out, denyJSON) {
		t.Fatalf("got %s, want %s", string(out), string(denyJSON))
	}
	if len(runner.calls) != 1 {
		t.Fatalf("expected 1 call, got %d", len(runner.calls))
	}
	if !strings.Contains(runner.calls[0].argv[0], "dos-hook") {
		t.Fatalf("expected native binary in argv[0], got %v", runner.calls[0].argv)
	}
}

func TestNativeDelegateFallsBackToPython(t *testing.T) {
	native := filepath.Join("tools", ".bin", "dos-hook")
	runner := newFakeRunner(
		reply{exitCode: 3, stdout: nil},
		reply{exitCode: 0, stdout: denyJSON},
	)
	out, err := RunHook("posttool", ".", testPayload, native, runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !bytes.Equal(out, denyJSON) {
		t.Fatalf("got %s, want %s", string(out), string(denyJSON))
	}
	if len(runner.calls) != 2 {
		t.Fatalf("expected 2 calls, got %d", len(runner.calls))
	}
	if !bytes.Equal(runner.calls[0].stdin, testPayload) {
		t.Fatalf("first call stdin mismatch")
	}
	if !bytes.Equal(runner.calls[1].stdin, testPayload) {
		t.Fatalf("second call stdin mismatch")
	}
	cmdLine := strings.Join(runner.calls[1].argv, " ")
	if !strings.Contains(cmdLine, "dos.cli") {
		t.Fatalf("expected dos.cli in fallback argv, got %s", cmdLine)
	}
}

func TestNativeAbsentFallsBackToPython(t *testing.T) {
	runner := newFakeRunner(reply{exitCode: 0, stdout: []byte("ok")})
	out, err := RunHook("pretool", ".", testPayload, "", runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if string(out) != "ok" {
		t.Fatalf("got %s, want ok", string(out))
	}
	if len(runner.calls) != 1 {
		t.Fatalf("expected 1 call, got %d", len(runner.calls))
	}
	cmdLine := strings.Join(runner.calls[0].argv, " ")
	if !strings.Contains(cmdLine, "dos.cli") {
		t.Fatalf("expected dos.cli in argv, got %s", cmdLine)
	}
}

func TestPythonExitCodeCoercedToZero(t *testing.T) {
	runner := newFakeRunner(reply{exitCode: 2, stdout: nil})
	out, err := RunHook("pretool", ".", testPayload, "", runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(out) != 0 {
		t.Fatalf("expected empty stdout, got %s", string(out))
	}

	var stdout bytes.Buffer
	rc := Run(
		[]string{"pretool", "--workspace", "."},
		bytes.NewReader(testPayload),
		&stdout,
		io.Discard,
		map[string]string{"CLAUDE_PROJECT_DIR": "/no/such/root"},
		runner.Run,
	)
	if rc != 0 {
		t.Fatalf("expected rc 0, got %d", rc)
	}
	if stdout.Len() != 0 {
		t.Fatalf("expected empty stdout, got %s", stdout.String())
	}
}

func TestDenyPreservedFromNative(t *testing.T) {
	native := filepath.Join("tools", ".bin", "dos-hook")
	runner := newFakeRunner(reply{exitCode: 0, stdout: denyJSON})
	out, err := RunHook("pretool", ".", testPayload, native, runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !bytes.Equal(out, denyJSON) {
		t.Fatalf("got %s, want %s", string(out), string(denyJSON))
	}
}

func TestDenyPreservedFromPython(t *testing.T) {
	runner := newFakeRunner(reply{exitCode: 2, stdout: denyJSON})
	out, err := RunHook("pretool", ".", testPayload, "", runner.Run)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !bytes.Equal(out, denyJSON) {
		t.Fatalf("got %s, want %s", string(out), string(denyJSON))
	}
}

func TestMainSucceedsEvenOnPanicOrError(t *testing.T) {
	boomRunner := func(argv []string, stdin []byte) (int, []byte, error) {
		panic("backend blew up")
	}
	rc := Run(
		[]string{"pretool", "--workspace", "."},
		bytes.NewReader(testPayload),
		io.Discard,
		io.Discard,
		map[string]string{"CLAUDE_PROJECT_DIR": "/no/such/root"},
		boomRunner,
	)
	if rc != 0 {
		t.Fatalf("expected rc 0 on panic, got %d", rc)
	}

	errRunner := func(argv []string, stdin []byte) (int, []byte, error) {
		return 1, nil, errors.New("backend error")
	}
	rc = Main(
		[]string{"pretool", "--workspace", "."},
		bytes.NewReader(testPayload),
		io.Discard,
		io.Discard,
		map[string]string{"CLAUDE_PROJECT_DIR": "/no/such/root"},
		errRunner,
	)
	if rc != 0 {
		t.Fatalf("expected rc 0 on error, got %d", rc)
	}
}

func TestWorkspaceArgParsed(t *testing.T) {
	runner := newFakeRunner(reply{exitCode: 0, stdout: []byte("ok")})
	var stdout bytes.Buffer
	rc := Run(
		[]string{"pretool", "--workspace", "custom/project/dir"},
		bytes.NewReader(testPayload),
		&stdout,
		io.Discard,
		map[string]string{"CLAUDE_PROJECT_DIR": "/no/such/root"},
		runner.Run,
	)
	if rc != 0 {
		t.Fatalf("expected rc 0, got %d", rc)
	}
	if len(runner.calls) != 1 {
		t.Fatalf("expected 1 call, got %d", len(runner.calls))
	}
	cmdLine := strings.Join(runner.calls[0].argv, " ")
	if !strings.Contains(cmdLine, "--workspace custom/project/dir") {
		t.Fatalf("expected custom workspace in argv, got %s", cmdLine)
	}

	runnerDefault := newFakeRunner(reply{exitCode: 0, stdout: []byte("ok")})
	Run(
		[]string{"pretool"},
		bytes.NewReader(testPayload),
		&stdout,
		io.Discard,
		map[string]string{"CLAUDE_PROJECT_DIR": "/no/such/root"},
		runnerDefault.Run,
	)
	if len(runnerDefault.calls) != 1 {
		t.Fatalf("expected 1 call, got %d", len(runnerDefault.calls))
	}
	cmdLineDefault := strings.Join(runnerDefault.calls[0].argv, " ")
	if !strings.Contains(cmdLineDefault, "--workspace .") {
		t.Fatalf("expected default '.' workspace in argv, got %s", cmdLineDefault)
	}
}

func TestRepoRootPrefersClaudeProjectDir(t *testing.T) {
	root := RepoRoot(map[string]string{"CLAUDE_PROJECT_DIR": "/some/ws"})
	if filepath.ToSlash(root) != "/some/ws" {
		t.Fatalf("got %s, want /some/ws", root)
	}
}

func TestNativeBinaryResolutionPrefersProjectDir(t *testing.T) {
	bin := NativeBinary("/no/such/root")
	if bin != "" {
		t.Fatalf("expected empty string for nonexistent root, got %s", bin)
	}
}
